File: app/models/mission.rb
# == Schema Information
#
# Table name: missions
#
#  id           :integer          not null, primary key
#  leader_id    :integer          not null
#  title        :string           not null
#  description  :text             not null
#  user_limit   :integer
#  created_at   :datetime         not null
#  updated_at   :datetime         not null
#  compensation :integer
#  latitude     :float
#  longitude    :float
#  completed    :boolean          default(FALSE)
#  start_date   :string           not null
#  end_date     :string           not null
#

class Mission < ActiveRecord::Base
  validates :leader_id, :title, :description, :latitude, :longitude,
            :start_date, :end_date, presence: true
  validates :title, uniqueness: true

  belongs_to :leader, class_name: :User
  has_many :enlists, class_name: 'Enlist', dependent: :destroy
  has_many :enlisted_users, through: :enlists, source: :user
  has_many :follows, class_name: 'Follow', dependent: :destroy
  has_many :following_users, through: :follows, source: :user
  has_many :images, dependent: :destroy
  has_many :comments, as: :commentable
  has_many :ratings
  has_many :updates

  def self.filter_by(data_type, min, max, missions_set=false)
    missions_set ||= Mission.all

    missions_set.where(data_type => min..max)
  end

  def self.filter_by_avg_rating(min_rating, max_rating, missions_set=false)
    missions_set ||= Mission.all

    missions_set.select do |mission|
      mission.avg_rating >= min_rating.to_f &&
        mission.avg_rating <= max_rating.to_f
    end
  end

  def self.filter_by_title_fragment(fragment, missions_set=false)
    missions_set ||= Mission.all

    missions_set.where("title LIKE ?", "%" + fragment + "%")
  end

  def avg_rating
    return 0 if ratings.count == 0
    ratings.pluck(:value).inject(:+) / ratings.count
  end
end

File: app/assets/javascripts/models/mission.js
Nebulr.Models.Mission = Backbone.Model.extend({
  urlRoot: "api/missions",

  leader: function () {
    if (!this._leader) {
      this._leader = new Nebulr.Models.User();
    }
    return this._leader;
  },

  enlistedUsers: function () {
    if (!this._enlistedUsers) {
      this._enlistedUsers = new Nebulr.Collections.Users();
    }
    return this._enlistedUsers;
  },

  followingUsers: function () {
    if (!this._followingUsers) {
      this._followingUsers = new Nebulr.Collections.Users();
    }
    return this._followingUsers;
  },

  enlists: function () {
    if (!this._enlists) {
      this._enlists = new Nebulr.Collections.Enlists();
    }
    return this._enlists;
  },

  follows: function () {
    if (!this._follows) {
      this._follows = new Nebulr.Collections.Follows();
    }
    return this._follows;
  },

  images: function () {
    if (!this._images) {
      this._images = new Nebulr.Collections.Images();
    }
    return this._images;
  },

  comments: function () {
    if (!this._comments) {
      this._comments = new Nebulr.Collections.Comments();
    }
    return this._comments;
  },

  ratings: function () {
    if (!this._ratings) {
      this._ratings = new Nebulr.Collections.Ratings();
    }
    return this._ratings;
  },

  updates: function () {
    if (!this._updates) {
      this._updates = new Nebulr.Collections.Updates();
    }
    return this._updates;
  },

  parse: function (response) {
    if (response.leader) {
      this.leader().set(response.leader);
      delete response.leader_id;
    }
    if (response.enlisted_users) {
      this.enlistedUsers().set(response.enlisted_users);
      delete response.enlisted_users;
    }
    if (response.following_users) {
      this.followingUsers().set(response.following_users);
      delete response.following_users;
    }
    if (response.enlists) {
      this.enlists().set(response.enlists);
      delete response.enlists;
    }
    if (response.follows) {
      this.follows().set(response.follows);
      delete response.follows;
    }
    if (response.images) {
      this.images().set(response.images);
      delete response.images;
    }
    if (response.comments) {
      this.comments().set(response.comments, { parse: true });
      delete response.comments;
    }
    if (response.ratings) {
      this.ratings().set(response.ratings);
      delete response.ratings;
    }
    if (response.updates) {
      this.updates().set(response.updates);
      delete response.updates;
    }

    return response;
  }
});
